modules/hardware/common/shared-mem.nix

```diff
@@ -0,0 +1,232 @@
+      default = false;
+      description = mdDoc ''
+        Enables memsocket on host.
+      '';
+      apply =
+        enable:
+        if enable && !config.ghaf.shm.enable then
+          builtins.throw "Enable shared memory in order to use shm on host"
+        else
+          enable;
+    };
+    instancesCount = mkOption {
+      type = types.int;
+      default =
+        if config.ghaf.shm.enable_host then
+          (builtins.length config.ghaf.shm.vms_enabled) + 1
+        else
+          builtins.length config.ghaf.shm.vms_enabled;
+      description = mdDoc ''
+        Number of VMs entitled to use shared memory.
+      '';
+    };
+    serverSocketPath = mkOption {
+      type = types.path;
+      default = "/run/user/${builtins.toString config.ghaf.users.accounts.uid}/memsocket-server.sock";
+      description = mdDoc ''
+        Defines location of the listening socket.
+        It's used by waypipe as an output socket when running in server mode
+      '';
+    };
+    clientSocketPath = mkOption {
+      type = types.path;
+      default = "/run/user/${builtins.toString config.ghaf.users.accounts.uid}/memsocket-client.sock";
+      description = mdDoc ''
+        Defines location of the output socket. It's fed
+        with data coming from AppVMs.
+        It's used by waypipe as an input socket when running in client mode
+      '';
+    };
+    display = mkOption {
+      type = types.bool;
+      default = false;
+      description = "Display VMs using shared memory";
+      apply =
+        enable:
+        if enable && !config.ghaf.shm.enable then
+          builtins.throw "Enable shared memory in order to use shm display"
+        else
+          enable;
+    };
+  };
+
+  config.boot.kernelParams =
+    let
+      hugepagesz = "2M"; # valid values: "2M" and "1G", as kernel supports these huge pages' size
+      hugepages =
+        if hugepagesz == "2M" then config.ghaf.shm.memSize / 2 else config.ghaf.shm.memSize / 1024;
+    in
+    optionals config.ghaf.shm.enable [
+      "hugepagesz=${hugepagesz}"
+      "hugepages=${toString hugepages}"
+    ];
+  config.environment.systemPackages = optionals config.ghaf.shm.enable_host [
+    (pkgs.callPackage ../../../packages/memsocket { vms = config.ghaf.shm.instancesCount; })
+  ];
+
+  config.systemd.services.ivshmemsrv =
+    let
+      pidFilePath = "/tmp/ivshmem-server.pid";
+      ivShMemSrv =
+        let
+          vectors = toString (2 * config.ghaf.shm.instancesCount);
+        in
+        pkgs.writeShellScriptBin "ivshmemsrv" ''
+          chown microvm /dev/hugepages
+          chgrp kvm /dev/hugepages
+          if [ -S ${config.ghaf.shm.hostSocketPath} ]; then
+            echo Erasing ${config.ghaf.shm.hostSocketPath} ${pidFilePath}
+            rm -f ${config.ghaf.shm.hostSocketPath}
+          fi
+          ${pkgs.sudo}/sbin/sudo -u microvm -g kvm ${pkgs.qemu_kvm}/bin/ivshmem-server -p ${pidFilePath} -n ${vectors} -m /dev/hugepages/ -l ${
```

Comment:
Could you run the systemd service as `User = "microvm"; Group = "kvm";` instead of using sudo?

Reply:
Running the service under root is needed for setting ownership of /dev/hugepages - it must be done under `root`. Elevating privileges in a user-owned service using /run/wrappers/bin/sudo is not feasible since it prompts for a password. Additionally, using ${pkgs.sudo}/sbin/sudo is also not an option in this case.